File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int l = 0;
        int r = height.length - 1;
        int maxL = height[l];
        int maxR = height[r];
        int sum = 0;

        while (l < r) {
            if (height[l] <= height[r]) {
                if (maxL - height[l] > 0) {
                    sum += maxL - height[l];
                }
                l++;
                maxL = Math.max(maxL, height[l]);
                
            }
            else {
                if (maxR - height[r] > 0) {
                    sum += maxR - height[r];
                }
                r--;
                maxR = Math.max(maxR, height[r]);
            }

        }
        return sum;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int total{};

        int l = 0 ;
        int r = height.size() - 1;
        int maxL = height[l];
        int maxR = height[r];
        
        while (l < r) {
            int tmp;
            if (maxL <= maxR) {
                maxL = std::max(maxL, height[++l]);
                tmp = std::min(maxR, maxL) - height[l];
            } else {
                maxR = std::max(maxR, height[--r]);
                tmp = std::min(maxR, maxL) - height[r];
            }

            total += tmp > 0 ? tmp : 0;
        }


        return total;
    }
};